src/main/java/org/usfirst/frc4048/common/apriltags/TCPServer.java

```diff
@@ -0,0 +1,113 @@
+package org.usfirst.frc4048.common.apriltags;
+
+import edu.wpi.first.wpilibj.DriverStation;
+
+import java.io.BufferedInputStream;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.LinkedList;
+import java.util.Queue;
+import java.util.concurrent.LinkedTransferQueue;
+
+public abstract class TCPServer<T> extends Thread {
+
+    private final int port;
+    private ServerSocket serverSocket;
+    private Socket clientSocket;
+    private DataInputStream dataInputStream;
+    private BufferedInputStream bufferedInputStream;
+    private boolean running = false;
+    private final LinkedTransferQueue<T> readings = new LinkedTransferQueue<>();
+
+    /**
+     * @param port for network communication. Port must match between client and server (range 5800-5810 (rio limitations))
+     */
+    public TCPServer(int port) {
+        this.port = port;
+    }
+
+
+    @Override
+    public void run() {
+        try {
+            this.serverSocket = new ServerSocket(port);
+            this.clientSocket = null;
+        } catch (IOException e2) {
+            DriverStation.reportError("Could not start server on port " + port, true);
+            this.serverSocket = null;
+        }
+        running = true;
+        while (running) {
+            if (serverSocket == null) {
+                escape();
+                return;
```

Comment:
Not making a semantic change but cleaner to call `continue` instead of `return` here